Segundo Entregable

fgarcia1105 opened the pull request:
- Se Organizan los archivos en diferentes carpetas para mayor legibilidad
- Se opta por organizar código según MVVM
- Se agrega nuevo item en el Tabbar
- Se agrega una segunda pantalla que permitirá mostrar los diferentes filtros(Trending, Now Playing, popular, etc), mediante tableView y CollectionView
- Se crea una extension para la descarga de imágenes de manera asíncrona

![image](https://user-images.githubusercontent.com/57502345/192040996-0b0856e8-8509-4eb1-bb77-c567c99d3288.png)


![image](https://user-images.githubusercontent.com/57502345/192041066-84bf2e25-754b-4af5-b5c9-ab246edcf79f.png)

Commit:
- Se organizan los archivos para mayor legibilidad
	- se estructura de tal manera que se aplique un patron de arquitectura MVVM
	- Se cambia la manera de interactuar con la vista y controladores

fgarcia1105 committed:
Merge branch 'dev-fernando-garcia'

fgarcia1105 committed:
- Se Organizan los archivos en diferentes carpetas para mayor legibilidad
- Se opta por organizar codigo según MVVM
- Se agrega otro item en el Tabbar
- Se agrega una segunda pantalla que permitirá mostrar los diferentes filtros(Trending, Now Playing, popular, etc), mediante tableView y CollectionView
- Se crea una extension para la descarga de imágenes de manera asíncrona

JosueQuinonesRivera commented on BAZProjectC1/Extensions.swift:
```diff
@@ -0,0 +1,46 @@
+import UIKit
+
+// Mark: - For download image
+extension UIImageView {
+    func downloaded(from url: URL, contentMode mode: UIView.ContentMode = .scaleAspectFit){
+        contentMode = mode
+        URLSession.shared.dataTask(with: url) { (data, response, error) in
+            guard
+                let httpURLResponse = response as? HTTPURLResponse, httpURLResponse.statusCode == 200,
+                let mimeType = response?.mimeType, mimeType.hasPrefix("image"),
+                let data = data, error == nil,
+                let image = UIImage(data: data) else {return}
+            DispatchQueue.main.async() { [weak self] in
+                self?.image = image
+            }
+        }.resume()
+    }
+    
+    func downloaded(from link: String, contentMode mode: UIView.ContentMode = .scaleAspectFit) {
+        guard let url = URL(string: link) else { return }
+        downloaded(from: url, contentMode: mode)
+    }
+    
+}
+
+extension UIImage {
+      convenience init?(url: URL?) {
+        guard let url = url else { return nil }
+                
+        do {
+          self.init(data: try Data(contentsOf: url))
+        } catch {
+            debugPrint("Cannot load image from url: \(url) with error: \(error)")
+          return nil
+        }
+      }
+    
+}
```
Entiendo que esto es muy práctico, y puede quedarse asi para el entregable. Solo como comentario, esto rompe los principios SOLID debido a que el `UIImageView` ahora tiene 3 responsabilidades, settear la imagen, hacer un request a una URL, y castear esa data a una imagen valida. Si se quisiera tener un handlind error para cuando falle el request de la imagen, con la logica actual se tendria que meter tambien aqui y con eso añadir una responsabilidad más.

fgarcia1105 replied:
De momento lo dejé asi, ya que se pretende culminar los requerimientos y posteriormente refactorizar el codigo

JosueQuinonesRivera commented on BAZProjectC1/Main/Helpers/DataExample.swift:
```diff
@@ -0,0 +1,70 @@
+                            ,posterCollectionCell(posterImage: "img4.jpg", title: "Ghost")
+                        ]
+                     ]
+        )/*,
+        TableViewMovieCellModel(
+            sectionFilter: "Now Playing"
+            ,posters: [
+                        [
+                            posterCollectionCell(posterImage: "img1.jpg", title: "Spiderman")
+                            ,posterCollectionCell(posterImage: "img2.jpg", title: "Batman")
+                            ,posterCollectionCell(posterImage: "img3.jpg", title: "Ant Man")
+                            ,posterCollectionCell(posterImage: "img4.jpg", title: "Ghost")
+                        ]
+                     ]
+        ),
+        TableViewMovieCellModel(
+            sectionFilter: "Popular"
+            ,posters: [
+                        [
+                            posterCollectionCell(posterImage: "img1.jpg", title: "Spiderman")
```
Dejar código comentado no es buena práctica debido a que es altamente probable que no se use nunca y solo genera.

JosueQuinonesRivera commented on BAZProjectC1/Main/View/CollectionViewCell/CollectionViewCell.swift:
```diff
@@ -0,0 +1,34 @@
+    @IBOutlet weak var movieLabel: UILabel!
+    @IBOutlet weak var posterImage: UIImageView!
+    
+    override func awakeFromNib() {
+        super.awakeFromNib()
+        
+    }
```
Remueve este código, no esta siendo utilizado

JosueQuinonesRivera commented on BAZProjectC1/Main/View/CollectionViewCell/CollectionViewCell.swift:
```diff
@@ -0,0 +1,34 @@
+    }
+    
+    func setImage(urlEndpoint: String){
+//        let urlImage = "https://w7.pngwing.com/pngs/310/332/png-transparent-computer-icons-home-house-desktop-service-home-blue-logo-room.png"
```
Remover codigo comentado, si necesitas esta URL, guardala en una constante de clase o en otro lado para futuro.

fgarcia1105 replied:
Se remueve, ya que solo fue para prueba

JosueQuinonesRivera commented on BAZProjectC1/Main/View/CollectionViewCell/CollectionViewCell.swift:
```diff
@@ -0,0 +1,34 @@
+    
+    func setImage(urlEndpoint: String){
+//        let urlImage = "https://w7.pngwing.com/pngs/310/332/png-transparent-computer-icons-home-house-desktop-service-home-blue-logo-room.png"
+        let urlImage = "https://image.tmdb.org/t/p/w500\(urlEndpoint)"
+        if let url = URL(string: urlImage) {
+            self.posterImage.downloaded(from: url)
+        } else {
+            let bundle = Bundle(for: TableViewCell.self)
+            if let image = UIImage(named: "poster", in: bundle, compatibleWith: nil){
+                self.posterImage.image = image
+            }
+        }
+        
+    }
```
Si la intención es seguir MVVM de acuerdo a la descripción del PR. Este codigo rompe la arquitectura, el `UICollectionViewCell` es la parte de la vista, no deberia ser responsable de esta logica relacionada a la URL.

fgarcia1105 replied:
Se dejó asi, para su posterior refactor, con el fin de cumplir con los requerimientos iniciales

JosueQuinonesRivera commented on BAZProjectC1/Main/View/CollectionViewCell/CollectionViewCellModel.swift:
```diff
@@ -0,0 +1,22 @@
+// Modelo para mostrar los distintos filtros de las peliculas
+
+struct posterCollectionCell {
+    var posterImage: String
```
Siguiendo el comentario anterior, aqui estas guardando solo el string, podrias meter una variable calculada que arroja al menos la URL, o incluso la imagen utilizando un use case o helper estatico. Esto como plan B, lo mejor seria que el helper se utilize en el `ViewModel`

JosueQuinonesRivera commented on BAZProjectC1/Main/View/MoviesViewController.swift:
```diff
@@ -0,0 +1,84 @@
+    }
+    
+    func setupView(){
+        self.TableView.register(UINib.init(nibName: "TableViewCell", bundle: Bundle(for: TableViewCell.self)), forCellReuseIdentifier: "cell")
```
Utiliza un `reuseIdentifier` mas simbólico que solo 'cell', recuerda que Swift es un lenguaje cuya intension es ser lo más self-documented posible. Tambien esto aplica para tus clases de `TableViewCell`

JosueQuinonesRivera commented on BAZProjectC1/Main/View/MoviesViewController.swift:
```diff
@@ -0,0 +1,84 @@
+        let rowArray = postersMovieArray.movieObject[indexPath.section].posters[indexPath.row]
+        cell.updateCellWith(row: rowArray)
+        
+        //cell's delegate
```
Si lo agregas como `MARK` sera mas util en el futuro.

JosueQuinonesRivera replied:
`//MARK: - Cell delegation`

JosueQuinonesRivera commented on BAZProjectC1/Main/View/MoviesViewController.swift:
```diff
@@ -0,0 +1,84 @@
+            return UITableViewCell()
+        }
+        
+        let rowArray = postersMovieArray.movieObject[indexPath.section].posters[indexPath.row]
```
Aqui tambien se rompe MVVM, la vista no tendria que conocer el arreglo de peliculas, con esto se pierde el sentido de una arquitectura, esto funciona, pero no es MVVM

JosueQuinonesRivera commented on BAZProjectC1/Main/View/MoviesViewController.swift:
```diff
@@ -0,0 +1,84 @@
+        if let posterRow = didTappedInTableViewCell.rowWithPosters {
+            debugPrint(posterRow[index])
+            self.tappedCell = posterRow[index]
+             //Enviar a otra pantalla
```
Agrega un TODO
`// TODO: - Send to other screen`

Tambien recuerda que todo lo relacionado al codigo es mejor que vaya en ingles para mantener la coherencia del codigo, los metodos y clases ya se nombraron en ingles.

JosueQuinonesRivera commented on BAZProjectC1/Main/View/TableViewCell/TableViewCell.swift:
```diff
@@ -0,0 +1,85 @@
+}
+
+
+class TableViewCell: UITableViewCell {
```
Considera renombrarlo a algo mas detallado y propio del proyecto para que esto sea self-documented. Recuerda la escalabilidad :)

JosueQuinonesRivera commented on HeaderTable.swift:
```diff
@@ -0,0 +1,20 @@
+
+import UIKit
+
+class HeaderTable: UITableViewHeaderFooterView {
+
+    /*
+    // Only override draw() if you perform custom drawing.
+    // An empty implementation adversely affects performance during animation.
+    override func draw(_ rect: CGRect) {
+        // Drawing code
+    }
+    */
+
+}
```
Remueve el codigo comentado

fgarcia1105 replied:
Se remueve codigo comentado

JosueQuinonesRivera commented on BAZProjectC1/Main/View/TableViewCell/TableViewCell.swift:
```diff
@@ -0,0 +1,85 @@
+   
+    weak var cellDelegate: CollectionViewCellDelegate?
+    
+    var rowWithPosters: [posterCollectionCell]?
```
`posterCollectionCell` podria ser renombrado para seguir el standard a `PosterCollectionCell`

fgarcia1105 replied:
Se renombra de acuerdo a lo comentado

JosueQuinonesRivera commented on BAZProjectC1/Main/View/TableViewCell/TableViewCell.swift:
```diff
@@ -0,0 +1,85 @@
+    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        if let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "collectionviewcellid", for: indexPath) as? CollectionViewCell {
+            cell.movieLabel.text = self.rowWithPosters?[indexPath.item].title ?? ""
+            let urlEndPoint = self.rowWithPosters?[indexPath.item].posterImage ?? ""
```
lo mismo que para la tabla.

JosueQuinonesRivera commented on BAZProjectC1/Main/View/TrendingViewController.swift:
```diff
@@ -0,0 +1,69 @@
+    
+    
+    override func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
+        var config = UIListContentConfiguration.cell()
+       
+        let object = viewModel.movieDataArray[indexPath.row]
+
+        config.text = object.title
+        
+        let urlImage = "https://image.tmdb.org/t/p/w500\(object.posterPath)"
+        if let url = URL(string: urlImage) {
+            config.image = UIImage (url: url)
+        } else {
+            config.image = UIImage(named: "poster")
+        }
+        cell.contentConfiguration = config
```
Muy buen uso del `willDisplay` !
El unico tema aqui es que la vista esta siendo encargada del fetch de la imagen

JosueQuinonesRivera commented on BAZProjectC1/Main/View/TrendingViewController.swift:
```diff
@@ -0,0 +1,69 @@
+final class TrendingViewController: UITableViewController {
+
+    @IBOutlet weak var principalTabBar: UITabBarItem!
+    private var movies: [Movie] = []
```
Para probar la arquitectura, deberias remover esta variable, la vista deberia tener todo lo necesario a traves del `ViewModel`

fgarcia1105 replied:
Se remueve la variable y funciona

JosueQuinonesRivera commented on BAZProjectC1/Main/ViewModel/TrendingMovieViewModel.swift:
```diff
@@ -0,0 +1,49 @@
+        }
+    }
+    
+    private let apiKey: String = "f6cd5c1a9e6c6b965fdcab0fa6ddd38a"
+    private let urlBase = "https://api.themoviedb.org/3"
+    
+    func getMovies () {
+        let endPoint = "/trending/movie/day?api_key=\(apiKey)"
+        
+        guard let url = URL(string: urlBase + endPoint) else {
+            return
+        }
+        
+        URLSession.shared.dataTask(with: url) { (data, response, error) in
+            if error != nil {
+                debugPrint (error?.localizedDescription ?? "Error")
+                return
+            }
+            
+            guard let json = data else {return}
+            
+            //Serializar datos
+            do{
+                let result = try JSONDecoder().decode(Movie.self, from: json)
+                self.movieDataArray = result.results
+            } catch let error {
+                debugPrint("Error: \(error.localizedDescription)")
+            }
+        }.resume()
+    }
+    
+}
```
Tener esta logica aqui ya es mucho mejor!
Si quieres dar el siguiente paso, seria pasarlo a un helper, ya que el trabajo del ViewModel es hacer la conexion entre el modelo y lo que se debe pintar en la vista, no tal cual requests.

JosueQuinonesRivera reviewed (commented):
Para el entregable no hay problema con los comentarios relacionados a la arquitectura, pero si el PR dice que esto es MVVM, seria bueno mantenerlo en todo el proyecto asi como lo implementaste en uno de los modulos. Otra opcion podria ser indicar en la descripcion que solo ese modulo tiene MVVM implementado y el resto se migrara despues.

JosueQuinonesRivera commented:
Esta genial que hayas incluido las vistas en el PR! se ve super bien estructurado.
Puedes tambien manejarlo con etiquetas para darle el formato que quieres, por ejemplo:

`<img src="URL-de-la-imagen-despues-del-upload-a-Github" alt="map" width="300" height="620"/>`

esa etiqueta para imagenes les dara un formato mas legible al momento del review. Recuerda que todos los comentarios de Github includa la descripcion del PR son Markdown.

fgarcia1105 committed:
- Se renombre variables para mantener la un mismo estándar
	- Se quitan comentarios que no se utilizan
	- El modulo de mostrar trending de películas se pasó a la arquitectura MVVM, una vez concluido con los requerimientos iniciales se procederá refactorizar el codigo de los demás modelos para implementar la misma arquitectura.
